### ISSUE TITLE
mapreducedim attempts to compute f(zero(T)), but it may throw

### ISSUE THREAD

**Issue by simonster** [JuliaLang/julia#10533]
[This line](https://github.com/JuliaLang/julia/blob/39d8aef7c6288d3fef481d56d49e3b054d3512e8/base/reducedim.jl#L96) assumes that `f(zero(T))` is safe to call. This is not always the case:

```
julia> sum(x->sqrt(x-1), ones(5), 1)
ERROR: DomainError:
sqrt will only return a complex result if called with a complex argument.
try sqrt (complex(x))
 in sqrt at math.jl:133
 in anonymous at none:1
 in reducedim_init at reducedim.jl:96
 in sum at reducedim.jl:255
```

**Comment by timholy** [comment]
I suppose we could

**Comment by StefanKarpinski** [comment]
Probably better to check applicability since inference can reason about that.

**Comment by timholy** [comment]
We can check applicability, but `sqrt(-1)` isn't something inference can reason about. I guess we need both?

**Comment by simonster** [comment]
If `!isempty(A)`, maybe we just use `f(A[1])` [here](https://github.com/JuliaLang/julia/blob/39d8aef7c6288d3fef481d56d49e3b054d3512e8/base/reducedim.jl#L96)? Then we wouldn't have to compute the reduction twice when 0 isn't in the domain of `f`.

**Comment by JeffBezanson** [comment]
`mapreduce` here has the best current approach I think: https://github.com/JuliaLang/julia/blob/97b18707a8c04358980a819f5f3b0008b13d77c4/base/reduce.jl#L133

I can't believe we're computing the whole reduction twice in some cases. I don't know how I missed that. What's more, if `zero` is not applicable, the code still tries to do `zero(sum(f,A))`. Sure, it *might* be defined for the result of `sum` and not for `T`, but if we're bothering to check, why make that assumption?

Generally `method_exists` is a code smell. It's also irrelevant to this issue. While `reduce` has a little leeway in its definition, anything involving `map` is quite rigid: the argument function should only be called on elements of the passed collection.

**Comment by inkydragon** [comment]
Looks like it's fixed on 1.6.3?

```julia
julia> sum(x->sqrt(x-1), ones(5); init=1)
1.0

julia> versioninfo()
Julia Version 1.6.3
Commit ae8452a9e0 (2021-09-23 17:34 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
```

**Comment by brenhinkeller** [comment]
Yeah, seems to be fixed nowadays; I think this is safe to close but feel free to reopen if not!
```
julia> sum(x->sqrt(x-1), ones(5); init=1)
1.0

julia> sum(x->sqrt(x-1), ones(5))
0.0
```
